hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/AbstractRealtimeRecordReader.java

```diff
@@ -77,19 +74,17 @@ private boolean usesCustomPayload() {
   }
 
   /**
-   * Goes through the log files in reverse order and finds the schema from the last available data block. If not, falls
+   * Gets schema from HoodieTableMetaClient. If not, falls
    * back to the schema from the latest parquet file. Finally, sets the partition column and projection fields into the
    * job conf.
    */
-  private void init() throws IOException {
-    Schema schemaFromLogFile = LogReaderUtils.readLatestSchemaFromLogFiles(split.getBasePath(), split.getDeltaLogFiles(), jobConf);
-    if (schemaFromLogFile == null) {
-      writerSchema = InputSplitUtils.getBaseFileSchema((FileSplit)split, jobConf);
-      LOG.info("Writer Schema From Parquet => " + writerSchema.getFields());
-    } else {
-      writerSchema = schemaFromLogFile;
-      LOG.info("Writer Schema From Log => " + writerSchema.toString(true));
-    }
+  private void init() throws Exception {
+
+    HoodieTableMetaClient metaClient = HoodieTableMetaClient.builder().setConf(split.getPath().getFileSystem(jobConf).getConf()).setBasePath(split.getBasePath()).build();
+    TableSchemaResolver schemaUtil = new TableSchemaResolver(metaClient);
```

Comment:
If diff partitions have diff schemas, this would fail right? partition2's schema could be unrelated to partition1 schema. so, I am not sure on the fix. can you help me understand how the fix would help the use-case where different partitions have diff schemas.

Reply:
@nsivabalan  We found this problem a long time ago， this fix can really solve this problem，
 but i notice  an additional comment in this funcition：    // TODO(vc): In the future, the reader schema should be updated based on log files & be able
vc may have some extra considerations for this problem。

Reply:
@nsivabalan The idea behind this is that the TableSchemaResolver will provide us the latest schema. Some partitions might not have few columns that are present in main schema, but those will be set as NULL for that partition.

The issue with current logic is that it gets different schema for different partition which later conflict if multiple partitions are read at the same time.

@xiarixiaoyao As for the TODO mentioned, I'm not exactly sure about the complete context, but with this patch it nulls out the fields not present before (i.e. not in the partition schema but is there in latest schema)

Reply:
I think this will work. But i'm worried about performance. Are we going to initialize metaclient for each split? If yes, then maybe we can cache metaclient at a table level.

Reply:
yes. but we also need to be mindful of accessing Metaclient from withiin log record reader. layering does not sit well. why would log record reader instantiate or use Metaclient. may be we can pass in SchemaResolver only from upper layer. I mean, cache at the table/split level and pass that into AbstractLogRecordReaders.

Reply:
+1 on not using metaClient in record reader. 
@aditiwari01 What do you think of reafactoring as suggested above?

Reply:
It makes sense toh have this metaclient come from upper layer. Let me explore the code a bit more.

Reply:
@aditiwari01 : do you happened to have any updates here please. We are looking to get this in for 0.11. would appreciate if you can follow up on the feedback.

Reply:
Would it be fine if I get the schema from Metaclient in InputFormat (HoodieCombineHiveInputFormat) once and pass the schema to all the record readers from there?

Reply:
I checked the latest master and code has evolved due to key dedup for metadata and other requirements. So, already we have metaClient in AbstractHoodieLogRecordReader. Can you rebase w/ latest master and revisit the actual fix. may be the fix has to be done at higher layers, bcoz, I see the schema resolution is not done in this class (w/ latest master)